url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/network/datacenter/NetworkDatacenter.java

```diff
@@ -0,0 +1,229 @@
+
+	public Map<Integer, Integer> HostToSwitchid;
+
+	public Map<Integer, Switch> Switchlist;
+
+	public Map<Integer, Integer> VmtoHostlist;
+
+	/**
+	 * Get list of all EdgeSwitches in the Datacenter network One can design similar functions for
+	 * other type of switches.
+	 * 
+	 */
+	public Map<Integer, Switch> getEdgeSwitch() {
+		Map<Integer, Switch> edgeswitch = new HashMap<Integer, Switch>();
+		for (Entry<Integer, Switch> es : Switchlist.entrySet()) {
+			if (es.getValue().level == NetworkConstants.EDGE_LEVEL) {
+				edgeswitch.put(es.getKey(), es.getValue());
+			}
+		}
+		return edgeswitch;
+
+	}
+
+	/**
+	 * Create the VM within the NetworkDatacenter. It can be directly accessed by Datacenter Broker
+	 * which manage allocation of Cloudlets.
+	 * 
+	 * 
+	 */
+	public boolean processVmCreateNetwork(Vm vm) {
+
+		boolean result = getVmAllocationPolicy().allocateHostForVm(vm);
+
+		if (result) {
+			VmToSwitchid.put(vm.getId(), ((NetworkHost) vm.getHost()).sw.getId());
+			VmtoHostlist.put(vm.getId(), vm.getHost().getId());
+			System.out.println(vm.getId() + " VM is created on " + vm.getHost().getId());
+			double amount = 0.0;
+			if (getDebts().containsKey(vm.getUserId())) {
+				amount = getDebts().get(vm.getUserId());
+			}
+			amount += getCharacteristics().getCostPerMem() * vm.getRam();
+			amount += getCharacteristics().getCostPerStorage() * vm.getSize();
+
+			getDebts().put(vm.getUserId(), amount);
+
+			getVmList().add(vm);
+
+			vm.updateVmProcessing(CloudSim.clock(), getVmAllocationPolicy().getHost(vm).getVmScheduler()
+					.getAllocatedMipsForVm(vm));
+		}
+		return result;
+	}
+
+	/**
+	 * Processes a Cloudlet submission.
+	 * 
+	 * @param ev a SimEvent object
+	 * @param ack an acknowledgement
+	 * 
+	 * @pre ev != null
+	 * @post $none
+	 */
+	@Override
+	protected void processCloudletSubmit(SimEvent ev, boolean ack) {
+		updateCloudletProcessing();
+
+		try {
+			// gets the Cloudlet object
+			Cloudlet cl = (Cloudlet) ev.getData();
+
+			// checks whether this Cloudlet has finished or not
+			if (cl.isFinished()) {
+				String name = CloudSim.getEntityName(cl.getUserId());
+				Log.printLine(getName() + ": Warning - Cloudlet #" + cl.getCloudletId() + " owned by " + name
+						+ " is already completed/finished.");
+				Log.printLine("Therefore, it is not being executed again");
+				Log.printLine();
+
+				// NOTE: If a Cloudlet has finished, then it won't be processed.
+				// So, if ack is required, this method sends back a result.
+				// If ack is not required, this method don't send back a result.
+				// Hence, this might cause CloudSim to be hanged since waiting
+				// for this Cloudlet back.
+				if (ack) {
+					int[] data = new int[3];
+					data[0] = getId();
+					data[1] = cl.getCloudletId();
+					data[2] = CloudSimTags.FALSE;
+
+					// unique tag = operation tag
+					int tag = CloudSimTags.CLOUDLET_SUBMIT_ACK;
+					sendNow(cl.getUserId(), tag, data);
+				}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 168:177, 208:217.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.